libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,266 @@
+			)
+				return m_variables.at(_identifier.name);
+			else
+				return newRestrictedVariable();
+		},
+		[&](Literal const& _literal)
+		{
+			smtutil::Expression v = valueOfLiteral(_literal);
+			if (!m_useInt)
+				v = int2bv(v);
+			return v;
+		}
+	}, _expression);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeBuiltin(
+	evmasm::Instruction _instruction,
+	vector<yul::Expression> const& _arguments
+)
+{
+	vector<smtutil::Expression> arguments = applyMap(
+		_arguments,
+		[this](yul::Expression const& _expr) { return encodeExpression(_expr); }
+	);
+	switch (_instruction)
+	{
+	case evmasm::Instruction::ADD:
+		return wrap(arguments.at(0) + arguments.at(1));
+	case evmasm::Instruction::MUL:
+		return wrap(arguments.at(0) * arguments.at(1));
+	case evmasm::Instruction::SUB:
+		return wrap(arguments.at(0) - arguments.at(1));
+	case evmasm::Instruction::DIV:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			arguments.at(0) / arguments.at(1)
+		);
+	// TODO SDIV
+	case evmasm::Instruction::MOD:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			// TODO compute mod differently?
+			wrap(arguments.at(0) % arguments.at(1))
+		);
+	// TODO SMOD
+	case evmasm::Instruction::LT:
+		if (m_useInt)
+			return smtutil::Expression::ite(arguments.at(0) < arguments.at(1), constantValue(1), constantValue(0));
+		break;
+	case evmasm::Instruction::GT:
+		if (m_useInt)
+			return smtutil::Expression::ite(arguments.at(0) > arguments.at(1), constantValue(1), constantValue(0));
+		break;
+	case evmasm::Instruction::EQ:
+		return smtutil::Expression::ite(arguments.at(0) == arguments.at(1), constantValue(1), constantValue(0));
+	case evmasm::Instruction::ISZERO:
+		return smtutil::Expression::ite(arguments.at(0) == constantValue(0), constantValue(1), constantValue(0));
+	case evmasm::Instruction::AND:
+		if (m_useInt)
+			return smtutil::Expression::ite(
+				(arguments.at(0) == 0 || arguments.at(0) == 1) &&
+				(arguments.at(1) == 0 || arguments.at(1) == 1),
+				smtutil::Expression::ite(
+					arguments.at(0) == 1 && arguments.at(1) == 1,
+					constantValue(1),
+					constantValue(0)
+				),
+				bv2int(int2bv(arguments.at(0)) & int2bv(arguments.at(1)))
+			);
+		else
+			return arguments.at(0) & arguments.at(1);
+	case evmasm::Instruction::OR:
+		if (m_useInt)
+		{
+			return smtutil::Expression::ite(
+				(arguments.at(0) == 0 || arguments.at(0) == 1) &&
+				(arguments.at(1) == 0 || arguments.at(1) == 1),
+				smtutil::Expression::ite(
+					arguments.at(0) == 1 || arguments.at(1) == 1,
+					constantValue(1),
+					constantValue(0)
+				),
+				bv2int(int2bv(arguments.at(0)) | int2bv(arguments.at(1)))
+			);
+		}
+		else
+			return arguments.at(0) | arguments.at(1);
+	case evmasm::Instruction::NOT:
+		if (m_useInt)
+			return bv2int(~int2bv(arguments.at(0)));
+		else
+			return ~arguments.at(0);
+	// TODO SHL
```

Comment:
Ah, that's why the test below does not work.

Reply:
Yep